x/stakeibc/keeper/validator_selection.go

```diff
@@ -4,60 +4,255 @@ import (
 	}
-	for _, validator := range validators {
-		targetDelForVal := targetDelegation[validator.GetAddress()]
-		delegationDelta[validator.GetAddress()] = targetDelForVal.Sub(validator.Delegation)
+
+	// For each validator, store the amount that their delegation should change
+	delegationDeltas := []RebalanceValidatorDelegationChange{}
+	totalDelegationChange := sdkmath.ZeroInt()
+	for _, validator := range hostZone.Validators {
+		// Compare the target with either the current delegation
+		delegationChange := targetDelegation[validator.Address].Sub(validator.Delegation)
+
+		// Only include validators who's delegation should change
+		if !delegationChange.IsZero() {
+			delegationDeltas = append(delegationDeltas, RebalanceValidatorDelegationChange{
+				ValidatorAddress: validator.Address,
+				Delta:            delegationChange,
+			})
+			totalDelegationChange = totalDelegationChange.Add(delegationChange)
+		}
+		k.Logger(ctx).Info(fmt.Sprintf("Adding delegation: %v to validator: %s", delegationChange, validator.Address))
 	}
-	return delegationDelta, nil
+
+	// Sanity check that the sum of all the delegation change's is equal to 0
+	// (meaning the total delegation across ALL validators has not changed)
+	if !totalDelegationChange.IsZero() {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidRequest,
+			"non-zero net delegation change (%v) across validators during rebalancing", totalDelegationChange)
+	}
+
+	return delegationDeltas, nil
 }
 
 // This will get the target validator delegation for the given hostZone
 // such that the total validator delegation is equal to the finalDelegation
 // output key is ADDRESS not NAME
 func (k Keeper) GetTargetValAmtsForHostZone(ctx sdk.Context, hostZone types.HostZone, finalDelegation sdkmath.Int) (map[string]sdkmath.Int, error) {
 	// Confirm the expected delegation amount is greater than 0
-	if finalDelegation.Equal(sdkmath.ZeroInt()) {
-		k.Logger(ctx).Error(fmt.Sprintf("Cannot calculate target delegation if final amount is 0 %s", hostZone.ChainId))
-		return nil, types.ErrNoValidatorWeights
+	if finalDelegation.IsZero() {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidRequest,
+			"Cannot calculate target delegation if final amount is 0 %s", hostZone.ChainId)
 	}
 
 	// Sum the total weight across all validators
 	totalWeight := k.GetTotalValidatorWeight(hostZone)
 	if totalWeight == 0 {
-		k.Logger(ctx).Error(fmt.Sprintf("No non-zero validators found for host zone %s", hostZone.ChainId))
-		return nil, types.ErrNoValidatorWeights
+		return nil, errorsmod.Wrapf(types.ErrNoValidatorWeights,
+			"No non-zero validators found for host zone %s", hostZone.ChainId)
 	}
 	k.Logger(ctx).Info(utils.LogWithHostZone(hostZone.ChainId, "Total Validator Weight: %d", totalWeight))
 
 	// sort validators by weight ascending, this is inplace sorting!
+	// QUESTION: Now that we're sorting by weight + name, should we remove this?
```

Comment:
I think so, because the list is actually ordered now! We should test it in integration tests though (I think because we're testing for stochasticity, it might be challenging to test through unit tests? low confidence). I _think_ the only integration test change would be adding a third validator and then setting two of them to the same weight, but please double-check me on that

Honestly what a silly piece of code, reverse the list and then stable sort 😂

Reply:
yeah good call, we should definitely be pretty thorough in IG testing the rebalancing before we release this! Added a TODO in the spec

Reply:
Nice, thank you!